File: as1/p2/kcolor.erl
-module(kcolor).
-export([new_graph/0, add_edge/2, add_vertex/2, kcolor/2]).
-include_lib("proper/include/proper.hrl").
-include_lib("eunit/include/eunit.hrl").

-type vert() :: term().
-type edge() :: {vert(), vert()}.
-type graph() :: [{vert(), [vert()]}].

new_graph() -> newGraph().
add_edge(A, B) -> insertEdge(A, B).
add_vertex(A, B) -> insertVert(A, B).


%% to generate an empty graph
newGraph() -> [].

%% to insert an edge specified by {V1, V2} to graph G
-spec insertEdge(graph(), edge()) -> graph().
insertEdge(G, {V1, V2}) ->
        case V1 =:= V2 of 
            true -> insertVert(G,V1);
            false -> insertEdge(insertEdge(G,V1,V2),V2,V1)
    end.

insertEdge(G, V1, V2) -> 
    case G of 
        [] -> [{V1,[V2]}];
        [{V,Adj}|Ns] when V =:= V1 -> [{V,lists:usort([V2|Adj])}|Ns];
        [N1|Ns] -> [N1]++insertEdge(Ns,V1,V2)
    end.

%% to insert a vertex V1 to graph G
-spec insertVert(graph(), term()) -> graph().
insertVert(G, V1) ->
    case G of
        [] -> [{V1,[]}];
        [{V,_}|_] when V=:=V1 -> G;
        [N1|Ns] -> [N1] ++ insertVert(Ns, V1)
    end.

%% to find all adjacent vertices of the vertex V1 in graph G and return
%% them in a list
-spec getAdj(graph(), term()) -> [term()].
getAdj(G, V1) -> 
    case G of
        [] -> [];
        [{V,Adj}|_] when V=:=V1 -> Adj;
        [_|Ns] -> getAdj(Ns, V1)
    end.

%% to get all vertices in the graph G
-spec getVerts(graph()) -> [term()].
getVerts(G) -> 
    case G of
        [] -> [];
        % [{V,_} | Vs] -> [V]++getVerts(Vs)
        [{V, _}|Vs] -> [V|getVerts(Vs)]
    end.

%% to get all edges in the graph G 
%% Notice: a single edge will be returned in both directions
%%         edge 1 to 2 will be return as both {1,2} and {2,1}
-spec getEdges(graph()) -> [{term(), term()}].
getEdges(G) -> getEdges(G, []).
getEdges([], Edges) -> Edges;
getEdges([{V, Es}|Vs], Edges) ->
    getEdges(Vs, Edges++[{V,X}||X<-Es]).


%% ====================================================================
%% Property-based testing for graph data structure - 1
%% ====================================================================
%% to check if adajcent vertices are in pair: adajcent vertices of vertex 1
%% include vertex 2, at the same time, adajcent vertices of vertex 2 must
%% include vertex 1.
prop_pairwise_adj() -> 
    ?FORALL(L, random_graph(integer()), pairwise_test(L)).

%% generator to generate random graph with a random list
random_graph(T) -> 
    ?LET(L, list(T), random_insert([],L)).

random_insert(G, List) -> 
    case List of
        [] -> G;
        [V] -> insertVert(G,V);
        [V1,V2|Vs] -> case lists:nth(rand:uniform(2),[true,false]) of
                         true -> random_insert(insertEdge(G, {V1,V2}), Vs);
                         false -> random_insert(insertVert(G,V1), [V2|Vs])
                      end
    end.

%% randomly pick a vertex and check if it is adjacent to a random neighbor
%% of itself
pairwise_test(G) -> 
    case getVerts(G) of
        [] -> true;
        [V1|_] -> case getAdj(G,V1) of 
                      [] -> true;
                      [V2|_] -> lists:member(V1,getAdj(G,V2))
                  end
    end.

%% ====================================================================
%% Property-based testing for graph data structure - 2
%% ====================================================================
%% to check if vertices in all edges are a subset of the set of all 
%% vertices

prop_edge_verts() -> 
    ?FORALL(L, random_graph(integer()), edge_vert_test(L)).

edge_vert_test([]) -> true;
edge_vert_test(G) -> 
    Edges = getEdges(G),
    Verts = getVerts(G),
    EVerts = lists:foldr(
              fun({V1,V2},Acc) -> [V1,V2|Acc] end, [], Edges),
    length(lists:usort(Verts) -- lists:usort(EVerts)) >= 0.

%% ======================================================================
%% kcolor algorithm implementation

getColors(Num) -> 
    AllColors = [a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t,u,v,w,x,y,z], 
    AllColors -- lists:nthtail(Num, AllColors).

kcolor(G, Num) -> kcolor(G, G, [], getColors(Num)).

kcolor(_, [], Colored, _) -> Colored;
kcolor(G, [{V, _}|Vs], Colored, Colors) ->
    ValidColoring = [[{V, X}|Colored] || X<-Colors, isValid(G, Colored, {V,X})],
    case ValidColoring of 
        [] -> false;
        [Coloring1|Others] -> 
            FirstColoring = kcolor(G, Vs, Coloring1, Colors),
            case FirstColoring of
                false -> case Others of 
                             [] -> false;
                             _ -> kcolor(G, Vs, lists:nth(1, Others), Colors)
                         end; 
                _ -> lists:sort(FirstColoring)
            end
    end.

isValid(G, Colored, {V, Color}) -> 
   Adjs = getAdj(G, V), 
   lists:foldr(fun(X, Acc) -> 
            case getColor(X, Colored) =/= Color of 
               true -> true andalso Acc; 
               false -> false
            end
         end, true, Adjs).

getColor(_, []) -> null;
getColor(V, [{V1, Color}|Vs]) -> 
    case V =:= V1 of 
        true -> Color;
        false -> getColor(V, Vs)
    end.

%% ====================================================================
%% Property-based testing for kcolor implementation: 1
%% ====================================================================
%% test to check if the length of returned colors equals the number of
%% vertices
prop_color_length() -> 
    ?FORALL(L, random_graph(integer()), color_length_test(L)).

color_length_test([]) -> true;
color_length_test(G) ->
    Num = rand:uniform(26), 
    Colored = kcolor(G, Num),
    case Colored of 
       false -> true;
       _ -> length(Colored) =:= length(G)
    end.
%% ====================================================================
%% Property-based testing for kcolor implementation: 2
%% ====================================================================
%% test to check if the length of returned colors equals the number of
%% vertices

prop_adj_color() -> 
    ?FORALL(L, random_graph(integer()), adj_color_test(L)).

adj_color_test([]) -> true;
adj_color_test(G) -> 
    Num = rand:uniform(26), 
    % io:format("~p", [G]),
    % io:format("~p~n", [Num]),
    Colored = kcolor(G, Num), 
    case Colored of
        false -> true;
        _ -> checkEdges(getEdges(G), Colored)
    end.

checkEdges([], _) -> true;
checkEdges(_, []) -> false;
checkEdges([{V1, V2}|Es], Colored) -> 
    case getColor(V1, Colored) =:= getColor(V2, Colored) of
        true -> false;
        false -> checkEdges(Es, Colored)
    end.

File: as2/p2/Kcolor.hs
module Kcolor where

import Test.QuickCheck
import Data.List(nub, sort)


type Vert = Int
type Edge = (Vert, Vert)
type Graph = [(Vert, [Vert])]

newGraph :: Graph
newGraph = []

insertEdge :: Graph -> Edge -> Graph
insertEdge g (v1, v2) = 
    case v1 == v2 of 
        True -> insertVert g v1
        False -> insertDirEdge (insertDirEdge g (v1, v2)) (v2, v1)

insertDirEdge :: Graph -> Edge -> Graph
insertDirEdge g (v1, v2) =
    case g of 
        [] -> [(v1, [v2])]
        (v, adj):vs -> 
            if v == v1 then (v, nub (v2:adj)):vs
                       else (v, adj):(insertDirEdge vs (v1,v2))

insertVert :: Graph -> Vert -> Graph
insertVert g v = 
    case g of 
        [] -> [(v,[])]
        (v1,adj):vs -> 
            if v == v1 then g else (v1,adj):(insertVert vs v)

getAdj :: Graph -> Vert -> [Vert]
getAdj g v = 
    case g of 
        [] -> []
        (v1,adj):vs ->
            if v == v1 then adj else getAdj vs v

getVerts :: Graph -> [Vert]
getVerts g = 
    case g of 
        [] -> []
        (v,_):vs -> v:(getVerts vs)

getEdges :: Graph -> [Edge]
getEdges g = getEdgesAcc g []

getEdgesAcc [] acc = acc
getEdgesAcc ((v,adj):vs) acc = getEdgesAcc vs [(v,x)|x<-adj]++acc

-- Property-based test for graph data structure
-- 1. pairwise adjacent vertex testing

prop_pairwiseAdj l = 
    forAll (randomGraph (length l) l newGraph) $ pairwiseTest

randomGraph :: Int -> [Int] -> Graph -> Gen Graph
randomGraph n l g = do
        v1 <- elements l
        v2 <- elements l
        if n > 2 then randomGraph (n-1) l (insertEdge g (v1,v2))
                 else return g

pairwiseTest g = 
    case getVerts g of
        [] -> True
        (v1:_) -> case getAdj g v1 of
                    [] -> True
                    (v2:_) -> elem v1 (getAdj g v2)

-- Property-based test for graph data structure
-- 2. to check the set of vertices is the subset of 
-- set of all vertices
prop_edgeVert l = 
    forAll (randomGraph (length l) l newGraph) $ edgeVertTest

edgeVertTest [] = True
edgeVertTest g = 
    foldr (\x acc -> if elem x verts then acc else False) True edgeVerts
    where edges = getEdges g
          verts = nub $ getVerts g
          edgeVerts = nub $ foldr (\(v1,v2) acc -> v1:v2:acc) [] edges

-- Kcolor problem: implementation of brute-force algorithm
type Color = (Vert, Char)
kcolor :: Graph -> Int -> Maybe [Color]
kcolor g num = solve g (getColors num)

solve g colors = do
    let resultList = take 1 [x | x<- (getCombs g colors), isValid (getEdges g) x]
    case resultList of
        [] -> Nothing
        x:xs -> Just x

getCombs :: Graph -> [Char] -> [[Color]]
getCombs [] _ = []
getCombs _ [] = []
getCombs g colors = 
    foldl (\acc a -> [x:y|x<-a, y<-acc]) [[]] vertColor
    where 
        vertColor = foldl (\acc c -> [(c,x)|x<-colors]:acc) [] vs
        vs = getVerts g

getColors :: Int -> [Char]
getColors n = take n ['a'..'z']

isValid :: [Edge] -> [Color] -> Bool
isValid [] _ = True
isValid _ [] = False
isValid ((v1,v2):vs) colors = 
    if vertColor v1 colors == vertColor v2 colors
        then False else isValid vs colors
    where
        vertColor v1 ((v2,c):vs) = if v1 == v2 then c else vertColor v1 vs

-- property-based testing for kcolor implementation
-- 1. test to check if the length of returned colors equals
-- the number of all vertices
prop_colorLength l = 
    forAll (randomGraphWithNum (length l) l newGraph) $ colorLenTest 

randomGraphWithNum len l g = do
    graph <- randomGraph (2*len) l g
    num <- elements [0..(div len 2)]
    return (graph, num)

-- colorLenTest _ = True
colorLenTest ([], _) = True
colorLenTest (g, n) = 
    case kcolor g n of
        Just colored -> (length colored) == (length g)
        _ -> True

-- property-based testing for kcolor implementation
-- 2. to check if adjacent vertices are of different color 
prop_adjColor l = 
    forAll (randomGraphWithNum (length l) l newGraph) $ adjTest

adjTest ([],_) = True
adjTest (g, n) =
    case kcolor g n of
        Just colored -> foldl 
            (\acc (v1,v2) -> 
                if acc == False then acc
                else (vertColor v1 colored) /= (vertColor v2 colored))
                True (getEdges g) 
        _ -> True
    where vertColor v1 ((v2,c):vs) = if v1 == v2 then c else vertColor v1 vs
